content/blog/2020-09-25-efficient-path-profiling/index.md

```diff
@@ -0,0 +1,85 @@
+
+We can demonstrate this reasoning by trying it out on a real CFG:
+
+<img src="graph.png" style="width: 30%;">
+
+In this picture, we've already labeled the edges from $w\_1$, $w\_2$ and $w\_3$ as $0$. We can use the algorithm to figure out the values for $e\_1$, $e\_2$ and $e\_3$ and see why they provide us  with the property that we want. First, the algorithm will assign $e\_1$ to $0$ (it doesn't matter the order that we process the edges, but the first one processed will be $0$). Next, if $e\_2$ is selected to be processed, it'll be assigned $\text{NUMPATHS}(w\_1)=1$ per the algorithm. Finally, $e\_3$ will be assigned $\text{NUMPATHS}(w\_1) + \text{NUMPATHS}(w\_2)=2$. So the range of values (just one $0$) that the paths starting at $w\_2$ produced is shifted by 1, and the paths starting at $w\_3$ are shifted by 2, and accumulating these ranges into one range gives us $0..2$.
+
+### 2. Optimizing weight selection
+
+The weights selected in step 1 could have suboptimal overhead. They are not selected with minimizing the number of weights in mind. Ideally we'd want fewer edges to have weights, so we have to do less computation when we're profiling a program execution. The next step of the algorithm helps to alleviate this by using [some related work on program event counting][eventcounting]. At a high level, the algorithm finds a maximum cost spanning tree, where the cost is the sum of the weights found in the first part of the algorithm. Each chord $c$ of this spanning tree is then assigned a new weight computed from summing the weights from the first part of the algorithm found in a cycle that contains $c$, the source and the sink of the graph. This transformation also preserves the properties we wanted in the first part of the algorithm, that is, that each path starting at $v$ still maps to a unique number in the range $0..\text{NUMPATHS}(v)$.
+
+### 3. Instrumenting code
+
+The algorithm augments the code we're profiling so that when we run the code, we have logic that tracks which edges are traversed and which weights are added to our running sum. 
+
+### Regenerating paths
+
+Finally, the paper provides a straightforward algorithm to recover a path from a number.
+
+1. Initialize $R$ to have the generated path number, and $v$ to be the entry vertex of the CFG. Keep track of the starting vertex in $\text{PATH}$.
+2. Find edge $v \rightarrow w$ that has the largest $\text{VAL}(v \rightarrow w)$ that's less than $R$. Set $R$ to $R-\text{VAL}(v \rightarrow w)$, and set $v$ to $w$. Append $v$ to $\text{PATH}$.
+3. Stop when you end up at the sink. $\text{PATH}$ will contain a list of vertices.
+
+## Augmenting CFGs with cycles
+
+The algorithm described in this paper only works on directed acyclic graphs (DAGs). The paper describes a technique for modifying the CFG to transform it into a DAG but still provide accurate path profiling, which is crucial to profile programs with loops.
```

Comment:
Perhaps one more sentence of detail would be helpful here to explain how the CFG is transformed into a dag?

Reply:
Added some more details on the CFG augmentation.